src/loaders/ncbi_downloader/ncbi_downloader_helper.py

```diff
@@ -0,0 +1,94 @@
+import os
+import time
+from datetime import datetime
+
+import ftputil
+
+
+def _download_genome_file(
+        download_dir: str, 
+        gene_id: str, 
+        target_file_ext: list[str],
+        exclude_name_substring: list[str],
+        overwrite: bool = False
+) -> None:
+    # NCBI file structure: a delegated directory is used to store files for all versions of a genome
+    # e.g. File structure for RS_GCF_000968435.2 (https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/968/435/)
+    # genomes/all/GCF/000/968/435/ --> GCF_000968435.1_ASM96843v1/
+    #                              --> GCF_000968435.2_ASM96843v2/
+
+    ncbi_domain, user_name, password = 'ftp.ncbi.nlm.nih.gov', 'anonymous', 'anonymous@domain.com'
+
+    success, attempts, max_attempts = False, 0, 3
+    while attempts < max_attempts and not success:
+        try:
+            time.sleep(5 * attempts)
+            with ftputil.FTPHost(ncbi_domain, user_name, password) as host:
+
+                gene_dir = '/genomes/all/{}/{}/{}/{}/'.format(
+                    gene_id[0:3], gene_id[4:7], gene_id[7:10], gene_id[10:13])
+
+                host.chdir(gene_dir)
+                dir_list = host.listdir(host.curdir)
+                gene_dir_name = [i for i in dir_list if i.startswith(gene_id)][0]
+                host.chdir(gene_dir_name)
+                gene_file_list = host.listdir(host.curdir)
+
+                for gene_file_name in gene_file_list:
+                    # file has target extensions but doesn't contain exclude name substring
+                    if any([gene_file_name.endswith(ext) for ext in target_file_ext]) and all(
+                            [substring not in gene_file_name for substring in exclude_name_substring]):
+                        result_file_path = os.path.join(download_dir, gene_file_name)
+                        if overwrite or not os.path.exists(result_file_path):
+                            host.download(gene_file_name, result_file_path)
+
+            success = True
+        except Exception as e:
+            print(f'Error:\n{e}\nfrom attempt {attempts + 1}.\nTrying to rerun.')
+            attempts += 1
+
+    if not success:
+        raise ValueError(f'Download Failed for {gene_id} after {max_attempts} attempts!')
+
+
+def download_genome_files(
+        gene_ids: list[str],
+        target_file_ext: list[str],
+        exclude_name_substring: list[str],
+        result_dir: str,
```

Comment:
@Tianhao-Gu I've been flip flopping on whether this module should understand the `collections` file structure and just take a root dir argument or be more flexible with where it stores the output. 

It seems weird to me to effectively pass in the source ID (NCBI) and hardcode it in the GTDB specific code when this code knows exactly what the source is.

Reply:
hmmm I didn't catch what you mean. We need to form the result dir somewhere essentially.

Reply:
IOW do we leave this as is, where it's just any directory, or instead do we pass in the collections root dir and this code appends `sourcedir/NCBI` (or whatever). For the latter, then the calling code doesn't have to hardcode `NCBI` anywhere, which really it shouldn't - this NCBI downloading code should know that it's supposed to save stuff in the NCBI directory.

Reply:
I see. Yea. I think the latter makes sense.

Reply:
Updated!

Reply:
This code is still just taking any old result dir - did you maybe not push your latest commit(s)?

Reply:
I misunderstood the conversation and made the change [here](https://github.com/kbase/collections/blob/a4c61eaa0b1bb6d0caba947564ec3cd3b9f939cd/src/loaders/ncbi_downloader/gtdb.py#L241). But looking it now since we already created work_dir/result previously in the script, why cannot we just pass in? Also, is more general and handy for testing purpose. Just my thoughts

Reply:
it's basically the reason given in the OP - it's weird for another script to be telling the NCBI script that the source of the data is NCBI. The ncbi script knows that, and so should be able to handle it itself. Maybe the way to do is have a function in the ncbi helper like

```
get_work_dir(root: Path):
     return root / loader_common_names.SOURCE_DATA_DIR / loader_common_names.DEFAULT_ENV
```
and call that from gtdb

Reply:
Updated!

Reply:
The NCBI helper code needs to create the work dir as there's no guarantee it's created when the code is called

Reply:
ah, this function is get_work_dir, not make_work_dir. A work_dir will be created when _make_work_dir is called.

Reply:
That function is never called in the ncbi code.

Reply:
But we already had https://github.com/kbase/collections/blob/89402df9d134b2fc93c06fee645e1754a4e2e499/src/loaders/ncbi_downloader/ncbi_downloader_helper.py#L82

Reply:
Ah, I missed that. I was expecting the work dir to be created before the threads were fired off so there was only one attempt to create it rather than `n_threads`.

Still probably fine as is so :+1:

Reply:
I moved everything ato get_work_dir in ncbi helper. Updated!